app/services/apple_auth_service.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+class AppleAuthService < AuthServiceBase
+  class << self
+    def find_user_by_token(apple_id_token, apple_given_name = nil, apple_family_name = nil)
+      decoded_token = decode_token(apple_id_token)
+      validate_token!(apple_id_token, decoded_token)
+      validate_expected_fields!(decoded_token)
+
+      apple_id_token = remote_user_id(decoded_token)
+      user_atts = extract_user_atts(apple_id_token, decoded_token, apple_id_token)
+      user_atts["first_name"] = apple_given_name if apple_given_name.present?
+      user_atts["last_name"] = apple_family_name if apple_family_name.present?
+      setup_user(apple_id_token, user_atts)
+    rescue JSON::ParserError => e
+      raise FailedAuthentication, e.message
+    rescue JWT::DecodeError => e
+      raise FailedAuthentication, e.message
+    rescue AppleAuth::Conditions::JWTValidationError => e
+      raise FailedAuthentication, e.message
+    end
+
+    private
+
+    def expected_fields
+      %w[sub email iss aud]
+    end
+
+    def decode_token(apple_id_token)
+      AppleAuth::JWTDecoder.new(apple_id_token).call
+    end
+
+    def validate_token!(apple_id_token, decoded_token)
+      raise FailedAuthentication, "Sub is missing from payload" unless decoded_token["sub"]
+      AppleAuth::UserIdentity.new(decoded_token["sub"], apple_id_token).validate!
```

Comment:
@andrewroth I believe we actually want to validate the `aud` field here, and not the `sub`.

@levieggertcru any input?